### PR TITLE
Allow rebalancing primary shards on shared filesystems

### LINKED ISSUE #10469: allow relocating primary shards on shared filesystems without failing the shard
In the current implementation of replica shard when a primary shard is relocated to a different node the cluster enters red state for the duration of the shared storage handoff. That duration is proportional to the shard size. 
The desired behavior is that the cluster doesn't enter the red state at all. An acceptable constraint is that for the duration of the storage handoff no documents can be added to the shard.

Wojtek K. 
Microsoft

### PR DESCRIPTION
**Pull request by dakrone**
Instead of failing the Engine for a shared filesystem, this change
allows a "soft close" of the Engine, where only the IndexWriter is
closed so that the replica can open an IndexWriter using the same
filesystem directory/mount.

Fixes #10469

### CODE REVIEW

**Comment by dakrone** [comment]
@s1monw this PR is still missing a test where replica failure is simulated (like you asked for), but I wanted to vet the idea of a "soft-close" of the Engine, I had to make some changes after #10452 was merged to allow an Engine to be closed without closing the Translog, because closing the translog caused all future recovery to hang when `translog.snapshot()` is called.

Let me know what you think and I will work on adding more tests tomorrow.

**Comment by s1monw** [comment]
@dakrone I don't think we should do the sep close methods an all that kind of stuff. I'd rather just call sync instead of close on the translog. Especially given the fact that @bleskes works on larger refactorings on that end so all the changes here are likely only needed on 1.x

**Comment by dakrone** [comment]
@s1monw wouldn't that leave the translog in a never-closed state then? Or is the translog closed somewhere else? Does that just rely on the injector being closed to close it?

**Comment by s1monw** [comment]
@dakrone the translog is close later - the sync is the important part..

**Comment by s1monw** [comment]
@dakrone  I think this is good functionality wise but I think the implementation needs to be less intrusive ie. I think we should implement this as a subclass of the engine instead of adding all these settings and changing how the recovery handler works and calling back into it.  I took a quick step copying your test and adding this quick and dirty to the engien factory and I think it's cleaner what do you think about this https://github.com/s1monw/elasticsearch/commit/5fd56da1f4cd321975c4b3eea1d5e6df413d7448

**Comment by dakrone** [comment]
@s1monw I updated this to use the method that you came up with

File: src/test/java/org/elasticsearch/index/IndexWithShadowReplicasTests.java [hunk 1]
```diff
@@ -310,6 +322,176 @@ public void testPrimaryRelocation() throws Exception {
     }
 
     @Test
+    @TestLogging("_root:DEBUG,index:TRACE")
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/index/IndexWithShadowReplicasTests.java]: do we still need that?

File: src/test/java/org/elasticsearch/index/IndexWithShadowReplicasTests.java [hunk 2]
```diff
@@ -310,6 +322,176 @@ public void testPrimaryRelocation() throws Exception {
...
+        logger.info("--> performing query");
+        flushAndRefresh();
+
+        SearchResponse resp = client().prepareSearch(IDX).setQuery(matchAllQuery()).get();
+        assertHitCount(resp, counter.get());
+        assertHitCount(resp, numPhase1Docs + numPhase2Docs);
+    }
+
+    @Test
+    @TestLogging("_root:DEBUG,index:TRACE,indices.recovery:TRACE")
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/index/IndexWithShadowReplicasTests.java]: do we still need that?

File: src/main/java/org/elasticsearch/index/engine/InternalEngineFactory.java [hunk 3]
```diff
@@ -18,9 +18,54 @@
  */
 package org.elasticsearch.index.engine;
 
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.util.concurrent.ReleasableLock;
+
 public class InternalEngineFactory implements EngineFactory {
     @Override
     public Engine newReadWriteEngine(EngineConfig config, boolean skipTranslogRecovery) {
+        // On a shared filesystem, we need to handle recovery slightly
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/engine/InternalEngineFactory.java]: can we move this to it's own class? and then use the same pattern as all the other Mockengine classes?

**Comment by s1monw** [comment]
left minor comments LGTM otherwise

**Comment by bleskes** [comment]
@dakrone this is the on that only goes to 1.x, right? asking because it still has the 2.0.0 label on it...

**Comment by s1monw** [comment]
@bleskes the plan is to push this impl to 1.x and another impl to master based on your refactoring in https://github.com/elastic/elasticsearch/pull/10624 makes sense? I removed the 2.0 label for now

**Comment by bleskes** [comment]
yeah, makes total sense, just checking because of the label. Thx,

File: src/test/java/org/elasticsearch/test/engine/MockSharedFSEngine.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+import org.elasticsearch.common.util.concurrent.ReleasableLock;
+import org.elasticsearch.index.engine.EngineConfig;
+import org.elasticsearch.index.engine.EngineException;
+import org.elasticsearch.index.engine.RecoveryEngineException;
+
+/**
+ * TODO: document me!
+ */
+public class MockSharedFSEngine extends MockInternalEngine {
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/test/engine/MockSharedFSEngine.java]: I think this should be a subclass of `SharedFSEngine` and use the new `MockEngineSupport` from https://github.com/elastic/elasticsearch/pull/10700

**Comment by s1monw** [comment]
sweet! LGTM

**Comment by dakrone** [comment]
This has been merged to 1.x only, I will rewrite and open a new PR once #10624 is merged into master, since it refactors much of the recovery process.

**Pull request closed without merging**